lib/AST/ASTContext.cpp

```diff
@@ -8191,6 +8191,225 @@ QualType ASTContext::mergeObjCGCQualifiers(QualType LHS, QualType RHS) {
   return QualType();
 }
 
+//===--------------------------------------------------------------------===//
+//                    Predicates For Checked C checked types
+//===--------------------------------------------------------------------===//
+
+static bool lessThan(bool Self, bool Other) {
+  return (Self != Other && !Self);
+}
+
+bool ASTContext::isAtLeastAsCheckedAs(QualType T1, QualType T2) const {
+  if (T1.isNull() || T2.isNull())
+    return true;
+
+  if (T1.getQualifiers() != T2.getQualifiers())
+    return false;
+
+  T1 = T1.getCanonicalType();
+  T2 = T2.getCanonicalType();
+
+  const Type *T1Type = T1.getTypePtr();
+  const Type *T2Type = T2.getTypePtr();
+
+  if (T1Type == T2Type)
+    return true;
+
+  Type::TypeClass T1TypeClass = T1->getTypeClass();
+  if (T1TypeClass != T2Type->getTypeClass())
```

Comment:
Would this, for example, consider `IncompleteArray` types and `ConstantArray` types separate?  Do we not consider `int [3]` to be at least as checked as `int[]`?